### ISSUE TITLE
Using a subclass of an stdlib dataclass as field fails validation

### ISSUE THREAD

**Issue by godinenbicicleta** [pydantic/pydantic#3466]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
# python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.8.2
            pydantic compiled: True
                 install path: /usr/local/lib/python3.9/site-packages/pydantic
               python version: 3.9.9 (main, Nov 17 2021, 16:20:40)  [GCC 10.2.1 20210110]
                     platform: Linux-5.10.47-linuxkit-x86_64-with-glibc2.31
     optional deps. installed: ['typing-extensions']
```

```py
from dataclasses import dataclass
from typing import TypeVar

from pydantic import BaseModel


@dataclass(frozen=True)
class Base:
    __annotations__ = {}

    def do_something():
        pass


BaseBounded = TypeVar("BaseBounded", bound=Base)


@dataclass(frozen=True)
class BaseWithParam(Base):
    a: str


class PydanticModel(BaseModel):
    u: BaseBounded


u = BaseWithParam(a="1")
assert isinstance(u, Base)
PydanticModel(u=u)

# ---------------------------------------------------------------------------
# ValidationError                           Traceback (most recent call last)
# <ipython-input-1-c2b627b7a10a> in <module>
#      27 u = BaseWithParam(a="1")
#     28 assert isinstance(u, Base)
# ---> 29 PydanticModel(u=u)
# 
# /usr/local/lib/python3.9/site-packages/pydantic/main.cpython-39-x86_64-linux-gnu.so in pydantic.main.BaseModel.__init__()
# 
# ValidationError: 1 validation error for PydanticModel
# u
#  instance of Base, tuple or dict expected (type=type_error.dataclass; class_name=Base)

```

## some additional info/context:

I think in previous versions (before dataclass support was added) this code would work just fine if the pydantic model had this config:
```
class Config:
   arbitrary_types_allowed = True
```

Then, the validation would only check `isinstance(u, Base)` and the validation would be successful. But now, pydantic will try to validate the dataclass and to do that it first converts it to a dict using the recursive method `dataclasses.asdict` which causes all information about specific subclasses to be lost

## possible solution
  
  I can make the code run without errors if I modify this function [_validate_dataclass](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/dataclasses.py#L40) to add an extra `elif` and use a non recursive version of `dataclasses.asdict` found here [python docs for dataclasses](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict)
  ```
def _validate_dataclass(cls: Type['DataclassT'], v: Any, config=None) -> 'DataclassT':
    if isinstance(v, cls):
        return v
    elif isinstance(v, (list, tuple)):
        return cls(*v)
    elif isinstance(v, dict):
        return cls(**v)
    # In nested dataclasses, v can be of type `dataclasses.dataclass`.
    # But to validate fields `cls` will be in fact a `pydantic.dataclasses.dataclass`,
    # which inherits directly from the class of `v`.
    elif is_builtin_dataclass(v) and cls.__bases__[0] is type(v):
        import dataclasses
        non_recursive_dict = dict((field.name, getattr(v, field_name)) for field in dataclasses.fields(v))
        return cls(**non_recursive_dict)
    elif is_builtin_dataclass(v) and isinstance(v, cls.__abses__[0]):
        import dataclasses
        non_recursive_dict = dict((field.name, getattr(v, field_name)) for field in dataclasses.fields(v))

        # note this is a hack, I hardcoded frozen=True to make it work
        target = dataclass(type(v), config=config, frozen=True)
        return target(**non_recursive_dict)
    else:
        raise DataclassTypeError(class_name=cls.__name__)
  ```
  
  to make sure config makes it to the function arguments, I also modified this line [dataclasses](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/dataclasses.py#L192) to be:
  ```
      from functools import partial
      cls.__validate__ = classmethod(partial(_validate_dataclass, config=config))  # type: ignore[assignment]
  ```
  
  This was a hack just to make the code run without validation errors, but if it makes sense I would be happy to try to make it a formal PR

**Comment by hb2638** [comment]
I just hit this problem today, thanks for reporting!

**Comment by PrettyWood** [comment]
It has been fixed in v1.10
Please report bugs on latest version only
